### PR TITLE
added logging

### PR DESCRIPTION
**Pull request by swdpcomputing**

### CODE REVIEW

**Review by ScenarioFactory — APPROVED** [review]

File: src/api/logging/log.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { createLogger } from '../common/helpers/logging/logger.js'
+import { LogCodes } from './log-codes.js'
+
+const logger = createLogger()
+
+const log = (logCode, params) => {
+  if (!logCode?.level || !logCode.event) {
```
Comment (Andrew-Folga) [inline src/api/logging/log.js:7]: Could we improve here for better error logging and clarity with:

`const log = (logConfig, params) => {
  const { level, event } = logConfig
  if (!level || !event) {
    throw new Error(`Invalid log configuration: Missing ${!level ? 'level' : 'event'}`)
  }

  const loggerOfType = getLoggerOfType(level)
  loggerOfType({
    event,
    ...params
  })
}`

File: src/api/logging/log.test.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    })
+    expect(mockInfoLogger).not.toHaveBeenCalled()
+    expect(mockErrorLogger).not.toHaveBeenCalled()
+  })
+
+  it('Should throw if logCode is null', () => {
+    expect(() => log(null, mockParams)).toThrow('Invalid logCode')
+  })
+
+  it('Should throw if logCode level is null', () => {
```
Comment (Andrew-Folga) [inline src/api/logging/log.test.js:118]: Should this not talk about `logCode event` being null instead of `logCode level` being null?

File: src/api/logging/log.test.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    expect(() =>
+      log(
+        {
+          level: 'info',
+          event: null
+        },
+        mockParams
+      )
+    ).toThrow('Invalid logCode')
```
Comment (Andrew-Folga) [inline src/api/logging/log.test.js:127]: Here it would be nicer if we provide finer scope of logging above in the actual code to test the finer scoped error msg (level vs event).

File: src/api/logging/log.test.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  it('Should throw if logCode event is null', () => {
+    expect(() =>
+      log(
+        {
+          level: null,
+          event: 'mock-info-event'
+        },
+        mockParams
+      )
+    ).toThrow('Invalid logCode')
```
Comment (Andrew-Folga) [inline src/api/logging/log.test.js:139]: Same as above.

File: src/api/logging/log.test.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    log(mockLogCodeInfo, nestedParams)
+
+    expect(mockInfoLogger).toHaveBeenCalledWith({
+      event: mockLogCodeInfo.event,
+      ...nestedParams
+    })
+  })
+})
```
Comment (Andrew-Folga) [inline src/api/logging/log.test.js:169]: Can we add tests that handle the following:

- Logging Behavior with Non-Object Parameters i.e. `log(mockLogCodeInfo, 'string param')` or 
`log(mockLogCodeError, 123)`
- Handling Edge Cases with Parameters - logging with deeply nested objects/arrays or very large objects i.e. `log(mockLogCodeInfo, { bigParam: new Array(1000).join('x') })`
- `undefined` as opposed to `null` - can we check we get the expected behaviours here?
-

**Review by Andrew-Folga — COMMENTED** [review]
Looks good overall - added a few comments for dicussion.

File: src/api/logging/log.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    throw new Error('Invalid logCode')
```
Comment (Andrew-Folga) [inline src/api/logging/log.js:8]: Should we not tie the logCode to LogCodes so we would enforce the use of valid LogCodes by validating logCode against LogCodes before calling the logger?

Something like:
```
const log = (logCode, params) => {
  if (!logCode?.level || !logCode.event || !LogCodes[logCode.level]?.[logCode.event]) {
    throw new Error('Invalid logCode')
  }

  getLoggerOfType(logCode.level)({
    event: logCode.event,
    ...params
  })
}
```

Or are we saying that `LogCodes` would live independent to `logCode` and they do not need to be strictly followed?

I thought it would be nice if we created the `LogCodes` to enforce it in the `log` function. WDYT?

File: src/api/logging/log.test.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        {
+          level: null,
+          event: 'mock-info-event'
+        },
+        mockParams
+      )
+    ).toThrow('Invalid logCode')
+  })
+
+  it('Should call the logger for a minimal valid logCode object', () => {
```
Comment (Andrew-Folga) [inline src/api/logging/log.test.js:142]: If we were to tie in the `event` with the actual `LogCodes` then this test should fail. Not sure what was the intention and whether we should not be tying it in/enforcing it but would be interested to have a chat about that.

File: src/api/logging/log.js (hunk 1 above)
Comment (swdpcomputing) [reply src/api/logging/log.js:7]: Done a version of this, destructured  level and event in the params, but didn't create a loggerOfType parameter as only called once.

File: src/api/logging/log.test.js (hunk 2 above)
Comment (swdpcomputing) [reply src/api/logging/log.test.js:118]: Wrong way round. Edited.

File: src/api/logging/log.test.js (hunk 3 above)
Comment (swdpcomputing) [reply src/api/logging/log.test.js:127]: Added.

File: src/api/logging/log.test.js (hunk 4 above)
Comment (swdpcomputing) [reply src/api/logging/log.test.js:139]: Added.

File: src/api/logging/log.test.js (hunk 5 above)
Comment (swdpcomputing) [reply src/api/logging/log.test.js:169]: Added.

File: src/api/logging/log.js (hunk 6 above)
Comment (swdpcomputing) [reply src/api/logging/log.js:8]: Added.

File: src/api/logging/log.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { createLogger } from '../common/helpers/logging/logger.js'
+import { LogCodes, LogCodeLevels } from './log-codes.js'
+
+const logger = createLogger()
+
+const log = ({ level, event } = {}, context = {}) => {
```
Comment (Andrew-Folga) [inline src/api/logging/log.js:6]: Can we add typings for this function? It also made me realise that we do not have jsdoc enforcing when we contribute which was a good thing. Here I was able to add this incorrect logging in the handler:

```
    log('fdasfds')
```

and it silently let me start the server up and failed in runtime which is a problem.

Comment (Andrew-Folga) [reply src/api/logging/log.js:6]: I guess we cannot have no static type safety with pure JS which is another matter but having typings where we can would help use it correctly I suppose? WDYT?

Something like:
```
/**
 * @typedef {'info' | 'error' | 'debug'} LogLevel
 */

/**
 * @typedef {object} LogCode
 * @property {LogLevel} level - The severity level of the log.
 * @property {string} event - The event identifier for the log.
 */

/**
 * Logs an event with the specified level and context.
 * @param {LogCode} logCode - The log code object containing level and event.
 * @param {object} [context] - Additional context to include in the log.
 * @throws {Error} Will throw an error if the logCode is missing or invalid.
 */
const log = ({ level, event } = {}, context = {}) => {
  validateLogParams({ level, event }, context)

  getLoggerOfType(level)({
    event,
    ...context
  })
}
```

Comment (Andrew-Folga) [reply src/api/logging/log.js:6]: Also would you be able to enable JSDoc enforcing like we have in `ffc-grants-eligibility-checker`?

Comment (swdpcomputing) [reply src/api/logging/log.js:19]: Added.
